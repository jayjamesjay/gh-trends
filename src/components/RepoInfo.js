/**
 * @module RepoInfo
 */

/**
 * Information about repository
 * @class
 */
class RepoInfo {
  /**
   * Creates a RepoInfo
   *
   * @param {string} nameWithOwner - name and author of repository
   * @param {string} url - url of the repository
   * @param {string} description - description
   * @param {number} stargazersCount - amount of stars
   * @param {string} language - programming language
   * @param {number} forks - amount of forks
   * @param {string} license - repository license
   */
  constructor(nameWithOwner, url, description, stargazersCount, language, forks, license) {
    this.nameWithOwner = nameWithOwner;
    this.url = url;
    this.description = description;
    this.stargazersCount = stargazersCount;
    this.language = language;
    this.forks = forks;
    this.license = license;
  }

  /**
   * Creates RepoInfo from Github API data
   *
   * @param {object} item - repository data from Github API
   * @returns {RepoInfo}
   */
  static fromGithubRes(item) {
    const { description, language, forks } = item;
    const license = item.license ? item.license.spdx_id : null;

    return new RepoInfo(
      item.full_name,
      item.html_url,
      description,
      item.stargazers_count,
      language,
      forks,
      license
    );
  }
}

export default RepoInfo;
