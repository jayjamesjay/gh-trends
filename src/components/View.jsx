import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { Content } from '../styles/Main';
import { ButtonMain } from '../styles/Button';
import InfoBlock from './InfoBlock';
import { RepoInfo, identicalItems } from './Data';

/**
 * @module View
 */

/**
 * Displays information about repositories from provided data
 *
 * @property {array} data - information about all repositories
 * @property {array} saved - information about repositories which has been saved
 * @property {function} save - saves repository
 * @returns {View}
 */
export default function View({ data, saved, save }) {
  return (
    <Content>
      {data.map(node => {
        const saving = saved.findIndex(item => identicalItems(item, node)) > -1;
        return <InfoBlock key={node.nameWithOwner} info={node} save={save} saved={saving} />;
      })}
    </Content>
  );
}

View.propTypes = {
  save: PropTypes.func.isRequired,
  saved: PropTypes.arrayOf(PropTypes.instanceOf(RepoInfo)).isRequired,
  data: PropTypes.arrayOf(PropTypes.instanceOf(RepoInfo)).isRequired
};

/**
 * View with button firing loadData function
 *
 * @property {function} loadData - loads information about more repositories
 * @property {array} data - information about all repositories
 * @property {array} saved - information about repositories which has been saved
 * @property {function} save - saves repository
 * @returns {ViewSingle}
 */
export function ViewSingle(props) {
  const { data, loadData } = props;

  return (
    <>
      <View {...props} />
      <footer>
        <ButtonMain visible={data.length > 0} onClick={loadData}>
          Show more
        </ButtonMain>
      </footer>
    </>
  );
}

ViewSingle.propTypes = {
  loadData: PropTypes.func.isRequired,
  data: PropTypes.arrayOf(PropTypes.instanceOf(RepoInfo)).isRequired
};

/**
 * View with button firing loadData function with specified id
 *
 * @property {function} loadData - loads information about more repositories
 * @property {string} id - id of this View
 * @property {array} data - information about all repositories
 * @property {array} saved - information about repositories which has been saved
 * @property {function} save - saves repository
 * @returns {ViewId}
 */
export function ViewId(props) {
  const { loadData, id } = props;
  const load = useCallback(() => loadData(id), [loadData, id]);
  return <ViewSingle {...props} loadData={load} />;
}

ViewId.propTypes = {
  loadData: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired
};
