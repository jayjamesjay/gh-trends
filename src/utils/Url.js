/**
 * @module Url
 */

/**
 * Adds language to query
 *
 * @param {string} base - current query
 * @param {string} lang - language which will be added
 * @returns {string}
 */
export function addLang(base, lang) {
  let result = base;
  let currLang = lang;

  if (lang === 'All Languages') {
    const langIdx = base.indexOf(' language:');

    if (langIdx > -1) {
      result = base.slice(0, langIdx);
    }
  } else if (!base.includes(`language:"${lang}"`)) {
    if (lang === 'C++') {
      currLang = 'cpp';
    }

    result = `${base} language:"${currLang}"`;
  }

  return result;
}

/**
 * Representaion of Url
 * @class
 */
export default class Url {
  /**
   * Creates a Url
   *
   * @param {string} api - base url
   * @returns {Url}
   */
  constructor(api) {
    this.api = api;
    this.queryStr = '';
    this.params = [];
  }

  /**
   * Adds more parameters to this Url
   *
   * @param {string} elem - part of Url
   * @returns {Url}
   */
  parts(...elem) {
    this.params = this.params.concat(...elem);
    return this;
  }

  /**
   * Adds query string to this Url
   *
   * @param {string} params - space separated list of parameters
   * @returns {Url}
   */
  query(params) {
    this.queryStr = params
      .split(' ')
      .map((elem) => elem.trim())
      .filter((elem) => elem !== '')
      .join('+');
    return this;
  }

  /**
   * Adds programming language to query
   *
   * @param {string} lang - programming language
   * @returns {Url}
   */
  lang(lang) {
    return this.query(addLang(this.queryStr, lang));
  }

  toString() {
    const query = `?q=${this.queryStr}`;
    return this.api + [query, ...this.params].join('&');
  }
}
