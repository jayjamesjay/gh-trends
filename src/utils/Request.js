import RepoInfoList from './RepoInfoList';
import Url from './Url';

/**
 * @module Request
 */

/** Default Github API url */
export const defApi = 'https://api.github.com/search/repositories';
/** Default sorting option */
export const sort = 'sort=stars';
/** Default amount of repositories per page */
export const defPerPage = 'per_page=6';

/**
 * Fetches data and returns promise which resolves with JSON response
 *
 * @param {string} url - request url
 * @returns {Promise}
 */
export async function getJSON(url) {
  const response = await fetch(url.toString());
  return response.json();
}

/** Set of loading states */
export const loadingState = {
  LOADED: 'loaded',
  INPROGRESS: 'in-progress',
  ERROR: 'error',
};

/**
 * Representaion of API request
 * @class
 */
export default class Request {
  /**
   * Creates a Request
   *
   * @param {string} api - base url for API
   * @param {string} perPage - amount of items per page
   * @returns {Request}
   */
  constructor(api = defApi, perPage = defPerPage) {
    this.api = api;
    this.perPage = perPage;
  }

  /**
   * Fetches data and updates dataContainer using dataContainerController.
   * Sends back information about current progress using loadingController.
   *
   * @param {function} loadingController - allows to update loading state
   * @param {RepoInfoList} dataContainer - stores currently loaded data
   * @param {function} dataContainerController - allows to update data
   * @param {string} query - space separated list of query parameters
   */
  async loadData(loadingController, dataContainer, dataContainerController, query) {
    loadingController(loadingState.INPROGRESS);

    try {
      const url = new Url(this.api).query(query).parts(this.perPage, `page=${dataContainer.page}`);
      const newList = new RepoInfoList(dataContainer.data, dataContainer.page);
      dataContainerController(newList);

      const { items } = await getJSON(url);
      newList.update(items);

      dataContainerController(newList);
      loadingController(loadingState.LOADED);
      // eslint-disable-next-line
    } catch (error) {
      loadingController(loadingState.ERROR);
    }
  }
}
