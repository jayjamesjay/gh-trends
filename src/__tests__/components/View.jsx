import React from 'react';
import 'jest-styled-components';
import { fireEvent, render, screen } from '@testing-library/react';

import View, { ViewSingle } from '../../components/View';
import { loadingState } from '../../utils/Request';
import RepoInfo from '../../utils/RepoInfo';

const initData = new Array(6);

for (let i = 0; i < initData.length; i += 1) {
  const curr = new RepoInfo(
    `jayjamesjay/gh-trends`,
    '',
    'Loading content for this website...',
    123,
    'JavaScript',
    321,
    'MIT',
  );
  curr.nameWithOwner += i;
  initData[i] = curr;
}

describe('<View />', () => {
  const func = () => {};
  const license = 'License: MIT';

  it(`renders with default style`, () => {
    const { asFragment } = render(<View data={[]} saved={[]} save={func} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders empty', () => {
    render(<View data={[]} saved={[]} save={func} />);
    expect(screen.queryAllByText(license)).toHaveLength(0);
  });

  it('renders with items', () => {
    render(<View data={initData} saved={[]} save={func} />);
    expect(screen.queryAllByText(license)).toHaveLength(6);
  });

  it('renders with saved', () => {
    render(<View data={initData} saved={initData.slice(0, 3)} save={func} />);
    expect(screen.queryAllByAltText('Remove from saved')).toHaveLength(3);
  });
});

describe('<ViewSingle />', () => {
  const id = 'view-1';
  let tempId;
  const func = () => {};
  const loading = loadingState.LOADED;
  const loadData = () => {
    tempId = id;
  };

  it(`renders with default style`, () => {
    const { asFragment } = render(
      <ViewSingle data={initData} saved={[]} save={func} loadData={func} loading={loading} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it(`renders empty and does not display Button`, () => {
    render(
      <ViewSingle
        data={[]}
        saved={[]}
        save={func}
        loadData={func}
        loading={loadingState.INPROGRESS}
      />,
    );
    expect(screen.queryByRole('button')).toBeNull();
  });

  it('renders with items and displays Button', () => {
    render(<ViewSingle data={initData} saved={[]} save={func} loadData={func} loading={loading} />);
    expect(screen.queryByText('Show more')).not.toBeNull();
  });

  it('fires loadData', () => {
    render(
      <ViewSingle data={initData} saved={[]} save={func} loadData={loadData} loading={loading} />,
    );
    fireEvent.click(screen.queryByText('Show more'));

    expect(tempId).toEqual(id);
  });
});
