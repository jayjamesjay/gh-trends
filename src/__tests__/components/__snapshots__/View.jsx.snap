// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<View /> renders with default style 1`] = `
<DocumentFragment>
  @media screen and (min-width:40rem) {
  .c0 {
    display: flex;
    flex-flow: row wrap;
    justify-content: space-around;
    margin: 0 1.5rem;
  }
}

<div
    class="c0"
  />
</DocumentFragment>
`;

exports[`<ViewSingle /> renders with default style 1`] = `
<DocumentFragment>
  .c5 {
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
  align-content: center;
  align-items: center;
}

.c6 {
  margin: auto 0 0 0;
}

.c7 {
  cursor: pointer;
  width: 2rem;
  height: 2rem;
  background: transparent;
  border: 0;
}

.c8 {
  margin: 0.5rem 0 0 0;
}

.c3 {
  display: block;
  font-size: 1.2rem;
  font-family: 'Noto Sans HK',sans-serif;
  color: #000;
  text-decoration: none;
  cursor: pointer;
}

.c1 {
  display: flex;
  margin: 2rem 1rem;
  min-height: 13rem;
  padding: 1.2rem 1.2rem 0.8rem 1.2rem;
  flex: 1 1 100%;
  flex-flow: column nowrap;
  background: linear-gradient( #fff 0%,#fff 0.3rem,#f1e05a 0.3rem,#f1e05a 0.6rem,#fff 0.6rem );
  box-shadow: 0px 3px 5px 1px rgba(217,217,217,0.82);
  border-radius: 0.5rem;
  text-align: left;
  word-break: break-word;
}

.c4 {
  font-weight: 400;
  font-size: 1.3rem;
  color: #767676;
}

.c2 {
  margin: 0.2rem 0;
  font-size: 1.7rem;
}

.c9 {
  width: 100%;
  height: auto;
  filter: invert(0%);
}

.c12 {
  display: inline-block;
  width: 1.5rem;
  height: 1.5rem;
  filter: invert(0%);
}

.c10 {
  margin: 0 0.5rem;
  padding: 0;
}

.c11 {
  display: flex;
  align-content: center;
  align-items: center;
  margin: 0 0.5rem;
  padding: 0;
}

@media screen and (min-width:40rem) {
  .c0 {
    display: flex;
    flex-flow: row wrap;
    justify-content: space-around;
    margin: 0 1.5rem;
  }
}

@media screen and (min-width:40rem) {
  .c1 {
    min-width: 16.5rem;
    margin: 1.5rem 1rem;
    flex: 1 1 27%;
  }
}

<div
    class="c0"
  >
    <article
      class="c1"
    >
      <h2
        class="c2"
      >
        <a
          class="c3"
          href=""
        >
          <span
            class="c4"
          >
            jayjamesjay
          </span>
           /gh-trends0
        </a>
      </h2>
      <p>
        Loading content for this website...
      </p>
      <div
        class="c5 c6"
      >
        <button
          class="c7 c8"
        >
          <img
            alt="Add to saved"
            class="c9"
            src="test-file-stub"
            title="Add to saved"
          />
        </button>
        <p
          class="c10"
        >
          Forks: 321
        </p>
        <p
          class="c10"
        >
          License: MIT
        </p>
        <p
          class="c11"
        >
          123
          <img
            alt="Stars"
            class="c12"
            src="test-file-stub"
          />
        </p>
        <p
          class="c10"
        >
          Language: JavaScript
        </p>
      </div>
    </article>
    <article
      class="c1"
    >
      <h2
        class="c2"
      >
        <a
          class="c3"
          href=""
        >
          <span
            class="c4"
          >
            jayjamesjay
          </span>
           /gh-trends1
        </a>
      </h2>
      <p>
        Loading content for this website...
      </p>
      <div
        class="c5 c6"
      >
        <button
          class="c7 c8"
        >
          <img
            alt="Add to saved"
            class="c9"
            src="test-file-stub"
            title="Add to saved"
          />
        </button>
        <p
          class="c10"
        >
          Forks: 321
        </p>
        <p
          class="c10"
        >
          License: MIT
        </p>
        <p
          class="c11"
        >
          123
          <img
            alt="Stars"
            class="c12"
            src="test-file-stub"
          />
        </p>
        <p
          class="c10"
        >
          Language: JavaScript
        </p>
      </div>
    </article>
    <article
      class="c1"
    >
      <h2
        class="c2"
      >
        <a
          class="c3"
          href=""
        >
          <span
            class="c4"
          >
            jayjamesjay
          </span>
           /gh-trends2
        </a>
      </h2>
      <p>
        Loading content for this website...
      </p>
      <div
        class="c5 c6"
      >
        <button
          class="c7 c8"
        >
          <img
            alt="Add to saved"
            class="c9"
            src="test-file-stub"
            title="Add to saved"
          />
        </button>
        <p
          class="c10"
        >
          Forks: 321
        </p>
        <p
          class="c10"
        >
          License: MIT
        </p>
        <p
          class="c11"
        >
          123
          <img
            alt="Stars"
            class="c12"
            src="test-file-stub"
          />
        </p>
        <p
          class="c10"
        >
          Language: JavaScript
        </p>
      </div>
    </article>
    <article
      class="c1"
    >
      <h2
        class="c2"
      >
        <a
          class="c3"
          href=""
        >
          <span
            class="c4"
          >
            jayjamesjay
          </span>
           /gh-trends3
        </a>
      </h2>
      <p>
        Loading content for this website...
      </p>
      <div
        class="c5 c6"
      >
        <button
          class="c7 c8"
        >
          <img
            alt="Add to saved"
            class="c9"
            src="test-file-stub"
            title="Add to saved"
          />
        </button>
        <p
          class="c10"
        >
          Forks: 321
        </p>
        <p
          class="c10"
        >
          License: MIT
        </p>
        <p
          class="c11"
        >
          123
          <img
            alt="Stars"
            class="c12"
            src="test-file-stub"
          />
        </p>
        <p
          class="c10"
        >
          Language: JavaScript
        </p>
      </div>
    </article>
    <article
      class="c1"
    >
      <h2
        class="c2"
      >
        <a
          class="c3"
          href=""
        >
          <span
            class="c4"
          >
            jayjamesjay
          </span>
           /gh-trends4
        </a>
      </h2>
      <p>
        Loading content for this website...
      </p>
      <div
        class="c5 c6"
      >
        <button
          class="c7 c8"
        >
          <img
            alt="Add to saved"
            class="c9"
            src="test-file-stub"
            title="Add to saved"
          />
        </button>
        <p
          class="c10"
        >
          Forks: 321
        </p>
        <p
          class="c10"
        >
          License: MIT
        </p>
        <p
          class="c11"
        >
          123
          <img
            alt="Stars"
            class="c12"
            src="test-file-stub"
          />
        </p>
        <p
          class="c10"
        >
          Language: JavaScript
        </p>
      </div>
    </article>
    <article
      class="c1"
    >
      <h2
        class="c2"
      >
        <a
          class="c3"
          href=""
        >
          <span
            class="c4"
          >
            jayjamesjay
          </span>
           /gh-trends5
        </a>
      </h2>
      <p>
        Loading content for this website...
      </p>
      <div
        class="c5 c6"
      >
        <button
          class="c7 c8"
        >
          <img
            alt="Add to saved"
            class="c9"
            src="test-file-stub"
            title="Add to saved"
          />
        </button>
        <p
          class="c10"
        >
          Forks: 321
        </p>
        <p
          class="c10"
        >
          License: MIT
        </p>
        <p
          class="c11"
        >
          123
          <img
            alt="Stars"
            class="c12"
            src="test-file-stub"
          />
        </p>
        <p
          class="c10"
        >
          Language: JavaScript
        </p>
      </div>
    </article>
  </div>
  .c0 {
  cursor: pointer;
  background: #fff;
  border: 0;
  border-radius: 0.5rem;
  font-size: 1.2rem;
  color: #000;
  text-transform: uppercase;
}

.c1 {
  margin: auto;
  padding: 1rem;
  box-shadow: 0px 3px 5px 1px rgba(217,217,217,0.82);
  width: auto;
  height: auto;
}

@media screen and (min-width:40rem) {

}

@media screen and (min-width:40rem) {

}

<footer>
    <button
      class="c0 c1"
    >
      Show more
    </button>
  </footer>
</DocumentFragment>
`;
